### ISSUE TITLE
Simplify experiment architecture around core research capabilities

### ISSUE THREAD

**Issue by xcz0** [xcz0/Eco-AutoDrive#97]
## Context

The current experiment tree exposes a number of study-level entrypoints under `src/eco_planner/experiments`, with matching configs and CLI registrations. This preserves the path by which the research developed, but it also means that several historical studies now participate in the architecture as if they were durable software capabilities.

For a personal research repository, that distinction may no longer be useful. The experiment layer could instead be organized around a smaller set of current scientific questions, while reusable execution, RL, reward, evaluation, and analysis mechanisms live in their natural modules.

This issue intentionally does not treat existing ADR constraints or backward compatibility as hard requirements. The aim is to clarify the direction of a refactor rather than prescribe an exact final tree up front.

## Motivation

The main concern is not simply the number of experiment files. A larger problem is that some experiment packages currently own mechanisms that appear reusable outside the study that introduced them, for example:

- fixed-batch rollout collection and persistence;
- reward reweighting, calibration, and component extraction;
- GAE / advantage / backward-only gradient diagnostics;
- PPO update diagnostics such as parameter deltas, post-update KL, policy-ratio movement, and policy probes;
- manual guidance intervention execution;
- statistics, matched comparisons, reporting, and figure generation.

Keeping such mechanisms under study-specific packages can make later experiments depend on earlier experiments rather than on lower-level capabilities. A useful refactor would try to reverse that dependency where it improves clarity and reuse.

## Working principles

The following principles are intended as guidance rather than architectural rules:

- `experiments/` should mainly describe **what scientific comparison is being run**: arms, sweeps, matched conditions, train/held-out splits, experiment-specific gates, and result selection.
- reusable mechanisms should generally live closer to the subsystem they operate on (`rl`, `evaluation`, `reward`, `analysis`, etc.);
- configs should continue to carry experiment semantics where that is easier to inspect and reproduce than embedding them in Python;
- analysis should preferably be reproducible from persisted raw/intermediate outputs rather than being inseparable from execution;
- historical issue/task/stage names should not, by themselves, justify permanent modules or CLI branches;
- the result should remain easy to navigate as a personal research project, so a generic workflow framework or excessive indirection would probably be counterproductive.

A practical test for experiment-layer code is:

> If the scientific condition changed but the underlying operation would still be useful, the operation may belong below `experiments/`.

For example, a lambda grid belongs to an experiment, while reweighting a collected rollout with another reward profile is likely a reusable reward operation.

## Research-facing responsibilities

The current studies seem to cluster around roughly the following scientific responsibilities. These boundaries are tentative and can be adjusted during the refactor.

### Closed-loop comparison

Keep a clear path for comparing:

- frozen Diffusion Planner;
- PPO fine-tuning without the energy term;
- PPO fine-tuning with the energy term;
- matched held-out evaluation over energy, driving behavior, route progress, and safety-related metrics.

The existing scalar-reward study is a natural source for this workflow, but issue/task identifiers and protocol guards introduced for a particular development stage do not necessarily need to survive as permanent architecture.

### Reward definition and calibration

One experiment-facing area could cover questions such as:

- reward component distributions;
- energy reward calibration and scale;
- lambda / reward-weight sensitivity;
- alternative energy representations;
- fixed-batch reward recomputation.

This is primarily about whether the objective being optimized has a sensible numerical and semantic definition.

### Credit assignment and objective diagnostics

It may be useful to separate reward definition from the question of how that reward affects PPO credit assignment. This area could cover:

- raw / centered / normalized advantage comparisons;
- GAE or critic-related ablations;
- objective decomposition;
- R0 / energy-only / intermediate-lambda comparisons;
- actor-gradient direction, norm, and attribution diagnostics.

The exact split between this and reward diagnostics should follow code cohesion rather than an imposed file count.

### Guidance experiments

The durable research questions appear to be approximately:

- **control authority**: does a guidance intervention produce a measurable longitudinal response above noise?
- **energy/behavior sweep**: how do guidance direction and strength affect energy use and driving behavior under matched scenarios?

These may remain one experiment module if the orchestration stays readable.

### PPO update diagnostics

The current `stability`, `effective-update`, and reproducibility work all contribute to the broader question:

> Does PPO produce a measurable, stable policy change without structural rollout collapse?

A future training-diagnostics workflow might therefore own only the experiment-specific search space and selection logic, while lower-level diagnostics are shared.

A small optimizer grid may remain useful, but multi-stage search, promotion, or pruning machinery should probably be retained only when it still answers a current research need.

## Candidate responsibilities to move below `experiments/`

The exact module names below are illustrative. The main point is the dependency direction.

### Fixed-batch rollout handling

Collection, loading, indexing, and concatenation of a reusable fixed rollout batch could plausibly move toward `rl/rollout/` rather than remain a reward experiment in its own right.

The experiment layer would then choose **which batch to collect and which conditions to compare**, while rollout code would implement **how a deterministic reusable batch is represented and collected**.

### Reward recomputation and reweighting

Operations such as:

- reweighting an episode with another reward profile;
- extracting reward-component matrices;
- constructing energy-only or weighted reward variants;
- applying reusable calibration transforms;

could plausibly move toward `rl/reward/`.

This would let reward and credit-assignment experiments consume reward operations without depending on another experiment package.

### Advantage and backward-only diagnostics

Reusable operations for:

- constructing diagnostic PPO batches;
- producing raw / centered / normalized advantage forms;
- computing actor gradients without optimizer steps;
- grouping gradients and comparing direction/norm;

could plausibly move toward `rl/optimization/` or another RL diagnostics module.

The experiment would then define which objective variants to compare rather than how backward-only probes are implemented.

### PPO update diagnostics

Training-run inspection currently includes several generally useful measurements, such as parameter movement, policy-ratio movement, gradient norms, post-update KL estimates, and checkpoint comparisons.

These may fit better under a shared RL diagnostics layer, with existing fixed-context policy probing remaining close to the policy/runtime code.

The experiment-specific part would be the thresholds, grid, and rule for selecting a useful update region.

### Guidance intervention execution

Manual guidance sweeps currently exercise a substantial amount of evaluation/runtime behavior: environment reset/step, fixed noise, inference, intervention application, episode lifecycle, and transition capture.

That execution path may be better represented as a reusable evaluation/intervention capability, leaving `experiments/guidance` to define intervention values, matched conditions, and the scientific comparison.

This should only be generalized as far as current guidance experiments actually need; a generic intervention framework is not a goal.

### Statistics and reporting

Descriptive statistics, matched differences, bootstrap summaries, gradient comparisons, result loading, and figure/report regeneration should continue to converge under `analysis/` where practical.

The experiment layer may decide which metrics are scientifically relevant, but it should not need to reimplement common statistical machinery.

## Possible experiment-layer granularity

After reusable mechanisms are moved down, the remaining experiment layer may naturally become much smaller. One plausible shape is roughly:

```text
src/eco_planner/experiments/
    comparison.py
    reward.py
    credit.py
    guidance.py
    training.py
```

An additional small protocol/comparison helper might make sense if several workflows genuinely share the same arm definitions or train/held-out split.

This is only a candidate granularity. The goal is not to enforce five files. A cohesive workflow may stay in one file; a file that starts accumulating unrelated execution, diagnostics, and reporting logic should probably be split or pushed downward instead.

Conversely, creating `experiments/common.py`, `experiments/utils.py`, or a new workflow framework solely to reach a tidy shape would likely move complexity rather than remove it.

## Configuration and CLI direction

Configs could remain organized around a small number of current research workflows, but the exact filenames do not need to mirror the Python layout.

Likewise, the CLI should expose research operations rather than internal module structure. A compact surface might eventually resemble:

```text
just exp compare train ...
just exp compare eval ...

just exp reward ...
just exp credit ...

just exp guidance authority ...
just exp guidance sweep ...

just exp training ...
```

The exact verbs should follow the workflows that remain after consolidation. `justfile` should preferably stay a thin convenience wrapper over a small Python CLI rather than becoming another experiment-definition layer.

## What should probably not define the architecture

The following may remain useful in archived configs, notes, results, or commit history, but they should not automatically imply long-lived packages or command hierarchies:

- issue/task identifiers (`issue81`, `Task F`, etc.);
- E-xxx experiment identifiers;
- stage A/B/C naming;
- top-N promotion pipelines introduced for one search procedure;
- thresholds introduced mainly while debugging one failed experiment;
- update-0 as a standalone protocol concept;
- machine/resource overrides;
- study-specific evidence/reference-directory abstractions;
- reproducibility as a separate experiment when ordinary multi-seed analysis is sufficient.

Some of these may still be worth keeping if they remain scientifically meaningful; the point is to avoid preserving them only because they once had dedicated code paths.

## Refactoring approach

A relatively low-risk sequence may be:

1. identify study-local helpers that are already reused, or are clearly independent of the study hypothesis;
2. move those mechanisms toward `rl`, `evaluation`, `reward`, or `analysis` with minimal semantic change;
3. reduce cross-imports between experiment packages;
4. consolidate historical studies only after their shared mechanisms have a lower-level home;
5. simplify configs and CLI names once the remaining experiment boundaries are clearer.

This ordering may avoid replacing many small experiment packages with a few large files that still contain the same underlying complexity.

## Non-goals

- Redesigning the core planner, simulator, PPO algorithm, or reward formulation.
- Introducing a generic workflow/orchestration framework.
- Preserving every historical entrypoint solely for compatibility.
- Forcing a particular file count or package tree before the code boundaries are tested.
- Generalizing an operation that is still genuinely specific to one current experiment.

## Indicators of a useful outcome

The refactor would be moving in the intended direction if most of the following become true:

- the experiment layer can be explained mainly in terms of current scientific questions rather than issue/task history;
- experiment modules mostly contain arms, sweeps, matched conditions, orchestration, and experiment-specific decisions;
- reward/credit experiments no longer need to import reusable mechanisms from another experiment package;
- fixed-batch collection can be reused without being a first-class research workflow;
- PPO update measurements can be reused by more than one training experiment without duplicating checkpoint/metric logic;
- guidance experiments can reuse an evaluation/intervention execution path rather than own a parallel evaluation engine;
- common statistics and report regeneration remain available from persisted results;
- the CLI remains small and research-oriented;
- the resulting source tree is easier to navigate without introducing a large new abstraction layer.

These are intended as design checks rather than strict acceptance criteria. The final module split should follow the code that remains after reusable mechanisms are separated from experiment composition.
